### PR TITLE
SolverRewards: Gnosis solvers score

### PR DESCRIPTION
**Pull request by sunce86**
Implements score calculation for gnosis solvers.
Score calculation is called in the post processing step as an additional optimisation step (which it is, kind of).

Note that for Quasimodo we don't define score computation parameters, therefore optimisation is skipped (because Quasimodo is now external solver and should submit the score independently).

Had to decouple `objective_value::Inputs` gas_amount and gas price so it can be reused.
Had to change trait `SettlementSimulating` to return gas_amount instead of bool.

### CODE REVIEW

File: crates/solver/src/settlement_post_processing/optimize_score.rs
Comment (MartinquaXD) [inline crates/solver/src/settlement_post_processing/optimize_score.rs:29]: Why would is 90% a more realistic gas amount here? Since we actually simulate the tx instead of using hardcoded gas estimates for each trade or interaction shouldn't this value be 100% correct?

File: crates/solver/src/solver/score_computation.rs
Comment (MartinquaXD) [inline crates/solver/src/solver/score_computation.rs:31]: I think either these need nicer names (e.g. `a -> gas_amount_factor`) or some comment which describes the revert probability function which relates the score computation inputs to these parameters.

File: crates/solver/src/solver/score_computation.rs
Comment (MartinquaXD) [inline crates/solver/src/solver/score_computation.rs:54]: Would be great to add a comment linking the code, simulation, whatever came up with these value if possible.

File: crates/solver/src/solver/score_computation.rs
Comment (MartinquaXD) [inline crates/solver/src/solver/score_computation.rs:55]: Would be nice to format all lines like: `Name,a,b,c,x;\n` to make it easier to read.

File: crates/solver/src/solver/score_computation.rs
Comment (MartinquaXD) [inline crates/solver/src/solver/score_computation.rs:86]: This (with more info + nicer wording) should probably be a doc-comment on the `score_parameters` field to generate a useful `--help` text for the CLI.

File: crates/solver/src/solver/score_computation.rs
Comment (MartinquaXD) [inline crates/solver/src/solver/score_computation.rs:46]: It's a bit surprising that all these values get negated. Is there a reason to do it like this instead of passing the already negated parameters via the CLI?

File: crates/solver/src/settlement_post_processing/mod.rs
Comment (MartinquaXD) [inline crates/solver/src/settlement_post_processing/mod.rs:9]: Originally didn't realize it but these modules can all be imported privately.

File: crates/solver/src/solver.rs
Comment (MartinquaXD) [inline crates/solver/src/solver.rs:386]: This information already gets printed on startup as part of the CLI arguments.

File: crates/solver/src/settlement_post_processing/mod.rs
Comment (MartinquaXD) [inline crates/solver/src/settlement_post_processing/mod.rs:143]: I would probably call it `compute_score()`. I agree that it kind of fits into the `PostProcessingPipeline` but just because the other steps are currently optimizations doesn't mean that this one also has to be called an optimization.

File: crates/solver/src/settlement_post_processing/optimize_score.rs
Comment (MartinquaXD) [inline crates/solver/src/settlement_post_processing/optimize_score.rs:30]: This can only happen if some random error happened during the simulation or if a new block got mined since the last optimization step where the settlement no longer works.
In both cases I would err on the safe side and set the `score` to 0 here to avoid having to execute a reverting solution.

File: crates/solver/src/settlement_post_processing/optimize_score.rs
Comment (sunce86) [reply crates/solver/src/settlement_post_processing/optimize_score.rs:29]: `eth_estimateGas` returns around 10-15% more gas than actually needed (because it's an _estimation_ function). The only way to calculate the gas exactly is to use Tenderly but IMO eth_estimateGas reduced by 10% already gives pretty good values. (Tenderly calls would take more time and increase our usage by 5-15 calls per auction).

Comment (sunce86) [reply crates/solver/src/settlement_post_processing/optimize_score.rs:29]: But I do agree additional explanation should be added to the comment.

File: crates/solver/src/solver/score_computation.rs
Comment (sunce86) [reply crates/solver/src/solver/score_computation.rs:55]: I tried but `fmt` messes everything up, and I did not find any solution other than turning it off for all strings which I didn't want to do.

File: crates/solver/src/solver/score_computation.rs
Comment (sunce86) [reply crates/solver/src/solver/score_computation.rs:46]: Hm, it's probably because the original formula that @harisang came up with make more sense using the `-`. TBH I did not get much into the formula itself.

Comment (harisang) [reply crates/solver/src/solver/score_computation.rs:46]: Fair point, @MartinquaXD . This is usually how logistic regression is formulated, and this is the only reason the parameters were computed like this. Of course what you propose is completely equivalent.

image

Comment (MartinquaXD) [reply crates/solver/src/solver/score_computation.rs:46]: Since it's usually done like this in math I think it's fine to leave it as is.

File: crates/solver/src/settlement_post_processing/optimize_score.rs
Comment (nlordell) [reply crates/solver/src/settlement_post_processing/optimize_score.rs:29]: > returns around 10-15% more gas than actually needed (because it's an estimation function).

It is also a product of needing it to be the total gas before refunds (an EVM requirement) so it will always be an upper bound on gas spent and not the actual gas spent.

File: crates/solver/src/solver/score_computation.rs
Comment (harisang) [inline crates/solver/src/solver/score_computation.rs:47]: This is actually the `success_probability`.

File: crates/solver/src/solver/score_computation.rs
Comment (harisang) [reply crates/solver/src/solver/score_computation.rs:54]: The code for collecting the data and training the model can be found here:
https://github.com/cowprotocol/risk_adjusted_rewards

The data for each solver can be found here.
https://drive.google.com/drive/u/1/folders/19yoL808qkp_os3BpLIYQahI3mQrNyx5T

File: crates/solver/src/solver/score_computation.rs
Comment (harisang) [inline crates/solver/src/solver/score_computation.rs:61]: Just confirmed that all values for all solvers are correct!

File: crates/solver/src/settlement_post_processing/mod.rs
Comment (sunce86) [reply crates/solver/src/settlement_post_processing/mod.rs]: Renamed to `estimate_gas`

File: crates/solver/src/settlement_post_processing/optimize_score.rs
Comment (sunce86) [reply crates/solver/src/settlement_post_processing/optimize_score.rs]: Added comment.

File: crates/solver/src/solver/score_computation.rs
Comment (sunce86) [reply crates/solver/src/solver/score_computation.rs]: Thanks @harisang added.

File: crates/solver/src/solver/score_computation.rs
Comment (sunce86) [reply crates/solver/src/solver/score_computation.rs]: `#[rustfmt::skip]` helped. fixed.

File: crates/solver/src/solver/score_computation.rs
Comment (sunce86) [reply crates/solver/src/solver/score_computation.rs]: Moved the comment to `score_parameters` and improved.

File: crates/solver/src/settlement_post_processing/mod.rs
Comment (sunce86) [reply crates/solver/src/settlement_post_processing/mod.rs]: Fixed

File: crates/solver/src/solver.rs
Comment (sunce86) [reply crates/solver/src/solver.rs]: removed

File: crates/solver/src/settlement_post_processing/mod.rs
Comment (sunce86) [reply crates/solver/src/settlement_post_processing/mod.rs]: renamed

File: crates/solver/src/settlement_post_processing/optimize_score.rs
Comment (sunce86) [reply crates/solver/src/settlement_post_processing/optimize_score.rs:29]: I kept is as is. The reasoning is that this component is expected to either improve the solution or do nothing. We do have a simulation again in the ranking phase that would cover failure. Setting score to 0 here is one side effect that I would prefer to avoid if possible (it would mean that it is the value provided by solver and harder to debug).

File: crates/solver/src/solver/score_computation.rs
Comment (sunce86) [reply crates/solver/src/solver/score_computation.rs]: Fixed.

File: crates/solver/src/solver/score_computation.rs
Comment (sunce86) [reply crates/solver/src/solver/score_computation.rs]: renamed. @harisang what should I call the parameter `x`?

Comment (harisang) [reply crates/solver/src/solver/score_computation.rs]: So, in the logistic regression literature it is called `intercept`. Not sure if that adds helps, but I guess it is more descriptive compared to `x`

Comment (sunce86) [reply crates/solver/src/solver/score_computation.rs]: Renamed

File: crates/solver/src/solver/score_computation.rs
Comment (nlordell) [inline crates/solver/src/solver/score_computation.rs:93]: I think it makes more sense to merge the default parameters with what is supplied via the command line. This makes overriding a single solver's score computation parameters easier.

**Review by nlordell — COMMENTED** [review]
Looks good to me.

One suggestion around parsing of the score parameters (i.e. we should have default values baked into code and have the CLI set overrides for each solver).

File: crates/solver/src/solver/score_computation.rs
Comment (sunce86) [reply crates/solver/src/solver/score_computation.rs:93]: Fixed.

File: crates/solver/src/settlement_post_processing/mod.rs
Comment (vkgnosis) [inline crates/solver/src/settlement_post_processing/mod.rs:12]: Somehow the imports got split. It should be modules first then all imports.

File: crates/solver/src/settlement_post_processing/optimize_score.rs
Comment (vkgnosis) [inline crates/solver/src/settlement_post_processing/optimize_score.rs:14]: From the perspective of this function it feels cleaner to take Settlement by reference and return the score. Up to you.

File: crates/solver/src/solver/score_computation.rs
Comment (vkgnosis) [inline crates/solver/src/solver/score_computation.rs:36]: To me it feels better to make the fields pub and not have a constructor in this case. Up to you.

File: crates/solver/src/solver/score_computation.rs
Comment (vkgnosis) [inline crates/solver/src/solver/score_computation.rs:161]: Maybe there should be at least one test to make sure the calculation is right. With an example from the solver team to check against.

**Review by vkgnosis — APPROVED** [review]

File: crates/solver/src/solver/score_computation.rs
Comment (nlordell) [inline crates/solver/src/solver/score_computation.rs:123]: My suggestion would be to, instead of parsing both, you would have a default:
```
fn default_score_parameters() -> HashMap<SolverType, ScoreCalculator> {
    // ...
}
```

And then have the parse function just override the default ones. It feals weird to encode default values as strings instead of actual Rust data structures.

**Review by nlordell — APPROVED** [review]

File: crates/solver/src/solver/score_computation.rs
Comment (nlordell) [reply crates/solver/src/solver/score_computation.rs:123]: Just a nit.

File: crates/solver/src/settlement_post_processing/mod.rs
Comment (sunce86) [reply crates/solver/src/settlement_post_processing/mod.rs]: Fixed.

File: crates/solver/src/settlement_post_processing/optimize_score.rs
Comment (sunce86) [reply crates/solver/src/settlement_post_processing/optimize_score.rs:14]: Yes, I tried to fit the function into the context of the place where it is used, but it does not make it really reusable. Fixed.

File: crates/solver/src/solver/score_computation.rs
Comment (sunce86) [reply crates/solver/src/solver/score_computation.rs]: Yes, no real need to be private here.

File: crates/solver/src/solver/score_computation.rs
Comment (sunce86) [reply crates/solver/src/solver/score_computation.rs:151]: YES! Was planning to test against one. I committed a test and now waiting for approval from @harisang

File: crates/solver/src/solver/score_computation.rs
Comment (fhenneke) [inline crates/solver/src/solver/score_computation.rs:43]: the scaling of `gas_amount` and `gas_price` is missing. otherwise the computation looks good.

(not sure if the following is valid rust)
```suggestion
        let exponent = self.intercept.neg()
            - self.gas_amount_factor * gas_amount * 1e-6
            - self.gas_price_factor * gas_price * 1e-1
            - self.nmb_orders_factor * nmb_orders as f64;
```

Comment (sunce86) [reply crates/solver/src/solver/score_computation.rs]: Thanks, added.

File: crates/solver/src/solver/score_computation.rs
Comment (sunce86) [reply crates/solver/src/solver/score_computation.rs:151]: Added test and confirmed the calculation with solver team.

**Pull request merged**
